core/tauri/src/resources/mod.rs

```diff
@@ -0,0 +1,159 @@
+/// to `rid`.
+pub(crate) type ResourceId = u32;
+
+/// Map-like data structure storing Tauri's resources (equivalent to file
+/// descriptors).
+///
+/// Provides basic methods for element access. A resource can be of any type.
+/// Different types of resources can be stored in the same map, and provided
+/// with a name for description.
+///
+/// Each resource is identified through a _resource ID (rid)_, which acts as
+/// the key in the map.
+#[derive(Default)]
+pub(crate) struct ResourceTable {
+  pub(crate) index: BTreeMap<ResourceId, Arc<dyn Resource>>,
+  pub(crate) next_rid: ResourceId,
+}
+
+impl ResourceTable {
+  /// Inserts resource into the resource table, which takes ownership of it.
+  ///
+  /// The resource type is erased at runtime and must be statically known
+  /// when retrieving it through `get()`.
+  ///
+  /// Returns a unique resource ID, which acts as a key for this resource.
+  pub(crate) fn add<T: Resource>(&mut self, resource: T) -> ResourceId {
+    self.add_arc(Arc::new(resource))
+  }
+
+  /// Inserts a `Arc`-wrapped resource into the resource table.
+  ///
+  /// The resource type is erased at runtime and must be statically known
+  /// when retrieving it through `get()`.
+  ///
+  /// Returns a unique resource ID, which acts as a key for this resource.
+  pub(crate) fn add_arc<T: Resource>(&mut self, resource: Arc<T>) -> ResourceId {
+    let resource = resource as Arc<dyn Resource>;
+    self.add_arc_dyn(resource)
+  }
+
+  pub(crate) fn add_arc_dyn(&mut self, resource: Arc<dyn Resource>) -> ResourceId {
+    let rid = self.next_rid;
+    let removed_resource = self.index.insert(rid, resource);
+    // TODO: add replace method if needed
+    assert!(removed_resource.is_none());
+    self.next_rid += 1;
+    rid
+  }
+
+  /// Returns a reference counted pointer to the resource of type `T` with the
+  /// given `rid`. If `rid` is not present or has a type different than `T`,
+  /// this function returns `None`.
+  pub(crate) fn get<T: Resource>(&self, rid: ResourceId) -> Result<Arc<T>, Error> {
+    self
+      .index
+      .get(&rid)
+      .and_then(|rc| rc.downcast_arc::<T>())
+      .map(Clone::clone)
+      .ok_or_else(|| Error::BadResourceId(rid))
+  }
+
+  // /// Removes a resource of type `T` from the resource table and returns it.
+  // /// If a resource with the given `rid` exists but its type does not match `T`,
+  // /// it is not removed from the resource table. Note that the resource's
+  // /// `close()` method is *not* called.
+  // ///
+  // /// Also note that there might be a case where
+  // /// the returned `Arc<T>` is referenced by other variables. That is, we cannot
+  // /// assume that `Arc::strong_count(&returned_arc)` is always equal to 1 on success.
+  // /// In particular, be really careful when you want to extract the inner value of
+  // /// type `T` from `Arc<T>`.
+  // pub(crate) fn take<T: Resource>(&mut self, rid: ResourceId) -> Result<Arc<T>, Error> {
```

Comment:
should we delete this?

Reply:
I left it there in case we ever need but I guess we can re-implement it